blog post draft (#23)
premAI-io/dev-portal: blog/2023-07-03-serve-llms-with-fastapi/index.md

tags: [llm, self-hosted, prem, open-source, fastapi, docker]
---

![Prem Banner](./banner.png)

<head>
  <meta name="twitter:image" content="./banner.png"/>
</head>

> In this tutorial, we are going to learn how to serve an LLM with FastAPI and Docker. For this tutorial I created a GitHub repostory with the code to serve `falcon-7b-instruct` already available [here](https://github.com/premAI-io/llm-fastapi-docker-template).

Reviewer: ```suggestion
In this tutorial, we are going to learn how to prepare a LLM with FastAPI and Docker to be able to be served as Prem Service on the Prem platform. For this tutorial, I created a GitHub repostory with the code to serve `falcon-7b-instruct` already available [here](https://github.com/premAI-io/llm-fastapi-docker-template).
```

> NOTE: in order to run Falcon 7B Instrcut model you will need a GPU with at least 16GiB of VRAM. You can use a [Paperspace Cloud](https://www.paperspace.com/gpu-cloud) virtual server or any other cloud provider or your own server with a NVIDIA GPU.

> ### Step 1: Setup Microservice Dependencies

Reviewer: ```suggestion
### Step 1: Setup the Python server
```

#### 1. Create `requirements.txt` file.

Create a `requirements.txt` file with the following dependencies:

```txt
fastapi==0.95.0
uvicorn==0.21.1
pytest==7.2.2
requests==2.28.2
tqdm==4.65.0
httpx==0.23.3
python-dotenv==1.0.0
tenacity==8.2.2
einops==0.6.1
sentencepiece==0.1.99
accelerate>=0.16.0,<1
xformers==0.0.20
```

> ### Step 2: Expose the LLM with FastAPI

Reviewer: ```suggestion
### Step 2: Expose the model with FastAPI
```

```python
import argparse
import os

import torch
import transformers
from tenacity import retry, stop_after_attempt, wait_fixed
from transformers import AutoTokenizer

parser = argparse.ArgumentParser()
parser.add_argument("--model", help="Model to download")
args = parser.parse_args()

print(f"Downloading model {args.model}")

@retry(stop=stop_after_attempt(3), wait=wait_fixed(5))
def download_model() -> None:
    _ = AutoTokenizer.from_pretrained(args.model, trust_remote_code=True)
    _ = transformers.pipeline(
        model=args.model,
        torch_dtype=torch.bfloat16,
        trust_remote_code=True,
        device_map=os.getenv("DEVICE", "auto"),
    )

download_model()

```

#### 2. Create a `Dockerfile`.

In order to serve an Hugging Face model we will use the official image from Huggingface which has all the necessary dependencies installed.

```dockerfile
FROM huggingface/transformers-pytorch-gpu:4.28.1

ARG MODEL_ID

WORKDIR /usr/src/app/

COPY requirements.txt ./

RUN pip install --no-cache-dir -r ./requirements.txt --upgrade pip

COPY download.py .

RUN python3 download.py --model $MODEL_ID

COPY . .

ENV MODEL_ID=$MODEL_ID

> CMD python3 main.py

Reviewer: does buildx honor the cache without multi-stage Dockerfile?
